File: backend/app/schemas/alerts.py
from pydantic import BaseModel, Field, ConfigDict
from uuid import UUID
from datetime import datetime
from typing import Optional, Literal, Any


# ============================================================================
# Alert Metadata Schemas
# ============================================================================

class TeamMemberJoinedMetadata(BaseModel):
    """Metadata for team member joined alerts."""
    user_id: UUID
    user_name: str
    user_email: str
    role: Optional[str] = None


class LowStockItemDetail(BaseModel):
    """Detail for a single low stock SKU."""
    sku_code: str
    sku_name: str
    available: int
    reorder_point: int


class LowStockMetadata(BaseModel):
    """Metadata for low stock alerts."""
    sku_codes: list[str]
    details: list[LowStockItemDetail]
    check_timestamp: datetime


# ============================================================================
# Alert Response Schemas
# ============================================================================

class AlertBase(BaseModel):
    """Base alert fields."""
    alert_type: Literal["team_member_joined", "low_stock"]
    severity: Literal["info", "warning", "critical"]
    title: str
    message: Optional[str] = None
    alert_metadata: Optional[dict[str, Any]] = None


class AlertResponse(AlertBase):
    """Alert response with read status for the requesting user."""
    model_config = ConfigDict(from_attributes=True)
    
    id: UUID
    aggregation_key: Optional[str] = None
    is_read: bool = Field(
        default=False,
        description="Whether the requesting user has read this alert"
    )


class UnreadCountResponse(BaseModel):
    """Unread alert count for sidebar badge."""
    count: int


class MarkReadResponse(BaseModel):
    """Response after marking alerts as read."""
    marked_count: int


# ============================================================================
# Internal Service Schemas
# ============================================================================

class LowStockItem(BaseModel):
    """Internal schema for low stock detection."""
    sku_code: str
    sku_name: str
    available: int
    reorder_point: int
    
    @property
    def severity(self) -> Literal["warning", "critical"]:
        """Determine severity based on available quantity."""
        if self.available <= 0:
            return "critical"
        return "warning"
    

File: frontend/src/lib/api/alerts.ts
import { protectedApiClient } from "./protected-client";

// Types
export interface AlertBase {
  alert_type: "team_member_joined" | "low_stock";
  severity: "info" | "warning" | "critical";
  title: string;
  message?: string | null;
  alert_metadata?: Record<string, any> | null;
}

export interface AlertResponse extends AlertBase {
  id: string;
  aggregation_key?: string | null;
  is_read: boolean;
  created_at: string;
}

export interface UnreadCountResponse {
  count: number;
}

export interface MarkReadResponse {
  marked_count: number;
}

export interface AlertsStatusResponse {
  alerts_enabled: boolean;
}

export interface AlertsParams {
  page?: number;
  size?: number;
  read?: "read" | "unread";
  alert_type?: "team_member_joined" | "low_stock";
}

export interface AlertsResponse {
  items: AlertResponse[];
  total: number;
  page: number;
  size: number;
  pages: number;
}

// API Functions

export async function getAlertsStatus(): Promise<AlertsStatusResponse> {
  return protectedApiClient<AlertsStatusResponse>("/alerts/status");
}

export async function getAlerts(
  params: AlertsParams = {}
): Promise<AlertsResponse> {
  const searchParams = new URLSearchParams();
  
  if (params.page) searchParams.append("page", String(params.page));
  if (params.size) searchParams.append("size", String(params.size));
  if (params.read) searchParams.append("read", params.read);
  if (params.alert_type) searchParams.append("alert_type", params.alert_type);

  const query = searchParams.toString();
  return protectedApiClient<AlertsResponse>(
    `/alerts${query ? `?${query}` : ""}`
  );
}

export async function getUnreadCount(): Promise<UnreadCountResponse> {
  return protectedApiClient<UnreadCountResponse>("/alerts/unread-count");
}

export async function markAlertAsRead(alertId: string): Promise<void> {
  await protectedApiClient(`/alerts/${alertId}/read`, {
    method: "POST",
  });
}

export async function markAllAlertsAsRead(): Promise<MarkReadResponse> {
  return protectedApiClient<MarkReadResponse>("/alerts/read-all", {
    method: "POST",
  });
}
